session-foundation/session-android
# Fix phone call from lock screen

### Description

Fix to prevent failure to display call notifications when the device is asleep or locked. Android 12 and above has tightened permissions with regard to starting services, in particular foreground services. As such, this patch wakes the device up in order to start the call service and display an incoming call notification, and may start a full-screen intent for the `WebRtcCallActivity` if necessary.

### Test steps:
With two devices running (Alice and Bob) - have Alice call Bob under the following conditions:
- With Bob having the app visible in the foreground,
- With Bob having the app running the background (i.e., dismissed but not closed),
- With Bob having force-closed the app (i.e., manually closed it from the back-stack) but the device is awake,
- With Bob having force-closed the app and then locked the device (in this case a full-screen intent will open the call activity, which will likely occur before the push notification arrives - but the push notification regarding the call _does_ (typically) arrive a few seconds later).

In each case Bob should receive an incoming call notification.

The scenario which has shown to be most problematic is if Bob has **force-closed the app**, and has the **lock on the app**, and has **locked their device**. In such a scenario, while most times we get the incoming call notification (but have to unlock the device twice - once for the device then again for the app),  sometimes we don't get the incoming call notification at all. Reasons for this discrepancy are unknown at the present time - but regardless of this issue, in my experience the call notification behaviour from this work significantly improves upon the current production behaviour.

Finally, it should be noted that Calls are in Beta and may not always successfully set up a call, so it may be possible to fail to initialise a call entirely, and this may not necessarily be caused by the changes in this branch.
